src/workerd/jsg/async-context.c++

```diff
@@ -0,0 +1,254 @@
+    : Scope(js.v8Isolate, resource) {}
+
+AsyncContextFrame::Scope::Scope(v8::Isolate* isolate, AsyncContextFrame& resource)
+    : isolate(IsolateBase::from(isolate)) {
+  this->isolate.pushAsyncFrame(resource);
+}
+
+AsyncContextFrame::Scope::~Scope() noexcept(false) {
+  isolate.popAsyncFrame();
+}
+
+AsyncContextFrame::StorageScope::StorageScope(
+    Lock& js,
+    StorageKey& key,
+    Value store)
+    : frame(AsyncContextFrame::create(js, nullptr, StorageEntry {
+        .key = kj::addRef(key),
+        .value = kj::mv(store)
+      })),
+      scope(js, *frame) {}
+
+void IsolateBase::pushAsyncFrame(AsyncContextFrame& next) {
+  asyncFrameStack.push_front(&next);
+}
+
+void IsolateBase::popAsyncFrame() {
+  asyncFrameStack.pop_front();
+  KJ_DASSERT(!asyncFrameStack.empty(), "the async context frame stack was corrupted");
+}
+
+void IsolateBase::promiseHook(v8::PromiseHookType type,
+                              v8::Local<v8::Promise> promise,
+                              v8::Local<v8::Value> parent) {
+  auto isolate = promise->GetIsolate();
+
+  // V8 will call the promise hook even while execution is terminating. In that
+  // case we don't want to do anything here.
+  if (isolate->IsExecutionTerminating() || isolate->IsDead()) {
+    return;
+  }
+
+  auto& js = Lock::from(isolate);
+
+  switch (type) {
+    case v8::PromiseHookType::kInit: {
+      // The kInit event is triggered by v8 when a deferred Promise is created. This
+      // includes all calls to `new Promise(...)`, `then()`, `catch()`, `finally()`,
+      // uses of `await ...`, `Promise.all()`, etc.
+      // Whenever a Promise is created, we associate it with the current AsyncContextFrame.
+      KJ_DASSERT(AsyncContextFrame::tryUnwrap(js, promise) == nullptr);
+      AsyncContextFrame::wrap(js, promise);
+      break;
+    }
+    case v8::PromiseHookType::kBefore: {
+      // The kBefore event is triggered immediately before a Promise continuation.
+      // We use it here to enter the AsyncContextFrame that was associated with the
+      // promise when it was created.
+      auto& frame = KJ_ASSERT_NONNULL(AsyncContextFrame::tryUnwrap(js, promise),
+          "the promise has no associated AsyncContextFrame");
+      IsolateBase::from(isolate).pushAsyncFrame(frame);
+      // We do not use AsyncContextFrame::Scope here because we do not exit the frame
+      // until the kAfter event fires.
+      break;
+    }
+    case v8::PromiseHookType::kAfter: {
+#ifdef KJ_DEBUG
+      auto& frame = KJ_ASSERT_NONNULL(AsyncContextFrame::tryUnwrap(js, promise),
+          "the promise has no associated AsyncContextFrame");
+      // The frame associated with the promise must be the current frame.
+      KJ_ASSERT(&frame == &AsyncContextFrame::current(js));
+#endif
+      IsolateBase::from(isolate).popAsyncFrame();
+      break;
+    }
+    case v8::PromiseHookType::kResolve: {
+      // This case is a bit different. As an optimization, it appears that v8 will skip
```

Comment:
Do we actually have to do anything for kResolve? I would guess that kBefore and kAfter never fire at all for such promises, so there's no need to attach the async context to them.

Reply:
Well, we definitely need to for `Promise.reject(...)` since the context needs to be attached and propagated to the `unhandledrejection` handler. For `Promise.resolve(...)` we can likely skip attaching the context since the promise isn't expected to be used anywhere else.

Reply:
Oh, hmm... I guess I find it a little surprising that it propagates to that callback.

This implies that, at least for the unhandled rejection event handler, the async context does NOT propagate from the point that called `addEventListener()`, but rather propagates from the place where the event was generated.

This worries me. Ideally I think async context should propagate from the call site of `addEventListener()`. Otherwise, there is a major difference between promise-style events vs. `addEventListener()`. Consider:

```
foo.onBar().then(event => {
  ...
});
```

vs.

```
foo.addEventListener("bar", event => {
  ...
})
```

Hopefully, these two behave basically the same as far as async local storage goes. However, it sounds like you're saying they might not: where promise continuations inherit the async context lexically (from the call site of .then()), it sounds like you're saying event handler callbacks may inherit the async context dynamically (from the place where the event was generated). That's quite different.

Personally, I would argue that these should be consistent, and, therefore, event handlers registered using `addEventListener()` should inherit lexically. But I guess that's not what Node does today?

What happens if a future standard decides that lexical inheritance is the way to go?

Reply:
Yeah I understand the confusion but this is currently the way Node.js works and it's the way the standard appears to be leaning.  It *could* change but based on recent conversations I would be surprised. 

Specifically, in Node.js, given the following:

```
const {
  AsyncLocalStorage,
} = require('async_hooks');

const als = new AsyncLocalStorage();

const p1 = als.run(123, () => Promise.reject('boom'));
const p2 = als.run(321, () => Promise.reject('bang'));

// This is Node.js equivalent unhandledrejection handler...
process.on('unhandledRejection', (...args) => {
  console.log(...args, als.getStore());
});
```

Each call to the `unhandledRejection` handler prints `123` and `321` respectively.